### PR TITLE
Allow Modal to place focus on first element within contents via new API

### LINKED ISSUE #54106: Modal itself receiving focus rather than first tabbable element
The standard `Modal` component defines a tabindex which forces it to be focused when mounted:

https://github.com/WordPress/gutenberg/blob/8e0fdcab990978929f2e452926aa5e66babe1c25/packages/components/src/modal/index.tsx#L232

As this is non-configurable it appears to be a deliberate decision but one that may be questionable from an a11y perspective as noted by @andrewhayward (see below).

Raising this Issue to discuss whether this attribute needs to be reconsidered or made configurable.

What is the rationale behind it and what arguments are there for/against removing it?

-------

For what it's worth, from an accessibility perspective at least, it's actually strongly advised that focus is moved to a control within the dialog, rather than keeping focus on the dialog itself. In this case, where the sole purpose of the dialog is to update the text value, the text field would be a good candidate to automatically receive focus here.

_Originally posted by @andrewhayward in https://github.com/WordPress/gutenberg/issues/53735#issuecomment-1700935928_

### PR DESCRIPTION
**Pull request by getdave**
## What?

Adds new API to `Modal` to allow focus to be placed on first element in the _content_ on mount.

Alternative to https://github.com/WordPress/gutenberg/pull/54296

Closes https://github.com/WordPress/gutenberg/issues/54106

## Why?

As discussed in https://github.com/WordPress/gutenberg/issues/54106 dialogs (including Modal) should ideally place focus on the first focusable element on mount. Currently all Modals find that the Close button is the first element in the DOM inside the Modal that is focusable and thus it receives focus. A11y feedback has been that this is unhelpful.

For further explanation as to how we arrived at this approach see

https://github.com/WordPress/gutenberg/pull/54296#discussion_r1325856634

## How?

Conditionally places the `focusOnMountRef` on the Modal _contents_ `<div>` when `firstContentElement` is passed as `focusOnMount` prop. This ensures the focus is attempted within the Modal's _contents_.

## Testing Instructions

- New Post
- Add some blocks and then Group them.
- Open List View
- Click options menu on Griup block and click `Rename`.
- See focus placed on `input` and not on `Close` button
- verify it's still possible to reach the `Close` button and constrained tabbing is still active.

## ✍️ Dev Note

The `Modal' component's `focusOnMount` prop has received an update, and it now accepts a new `"firstContentElement"` option. When setting `focusOnMount="firstContentElement"`, the `Modal` component will try to focus on the first tabbable element within the `Modal`'s content (ie. the markup passed via the `children` prop).

This is different from the pre-existing `"firstElement"` option, which causes focus to go to the first element anywhere within the `Modal`, including its header (usually the close button).

Note that it is the responsibility of the consumer to ensure that, when `focusOnMount="firstContentElement"`, there is at least one tabbable element within the `Modal`'s children.

### CODE REVIEW

File: packages/components/src/modal/index.tsx
Comment (getdave) [inline packages/components/src/modal/index.tsx:90]: I suppose this conditional may not be necessary as the string with coerce to `true`. Does that alter the behaviour 🤔

File: packages/components/src/modal/index.tsx
Comment (getdave) [inline packages/components/src/modal/index.tsx]: The hook requires that the parameter is a `ref` thus the `noopRef`.

File: packages/components/src/modal/index.tsx
Comment (getdave) [inline packages/components/src/modal/index.tsx]: I guess we could just not add the ref at all. Does passing `undefined` work with refs?

File: packages/components/src/modal/index.tsx
Comment (getdave) [inline packages/components/src/modal/index.tsx]: Trying to combine the focus ref with the `childrenContainerRef` results in errors due to the `useEffects` above. Any ways to avoid the extra div appreciated.

**Comment by getdave** [comment]
Given that 16.7 is the final release before WP 6.4 I'd really like to land this API so we can fix the Modal focus for the Block Rename feature 🙏

File: packages/components/src/modal/index.tsx
Comment (draganescu) [reply packages/components/src/modal/index.tsx:90]: So this change means:

- since the hook signature is `useFocusOnMount( focusOnMount = 'firstElement' )` 
- focusOnMount absent from props it means `firstElement`, but if it's present and has a value of `firstContentElement` then it means `firstElement`.
- otherwise if it's present and has any other value nothing changes.

File: packages/components/src/modal/index.tsx
Comment (stokesman) [reply packages/components/src/modal/index.tsx]: Can `useMergeRefs` not help? I was trying this and not seeing errors.
```jsx
						<div
							ref={
								useMergeRefs( [
									childrenContainerRef,
									focusOnMount === 'firstContentElement'
										? focusOnMountRef
										: null
							] }
						>
							{ children }
						</div>
```

File: packages/components/src/modal/index.tsx
Comment (stokesman) [reply packages/components/src/modal/index.tsx]: It does seem to be what the typing says yet have a look at how it's used in `useDialog`:
https://github.com/WordPress/gutenberg/blob/a2e4ac34eafe3602c5d4212b7573539fabd5c02a/packages/compose/src/hooks/use-dialog/index.ts#L85-L90

File: packages/components/src/modal/index.tsx
Comment (getdave) [reply packages/components/src/modal/index.tsx]: It does. Updated.

File: packages/components/src/modal/index.tsx
Comment (getdave) [reply packages/components/src/modal/index.tsx]: That's strange. I could have sworn I tried that and it didn't work. 

I've updated now and it does in fact work as you said. Thank you.

File: packages/components/src/modal/index.tsx
Comment (getdave) [reply packages/components/src/modal/index.tsx:90]: As you can see the original code was:

```jsx
const focusOnMountRef = useFocusOnMount( focusOnMount );
```

The new code means that if `firstContentElement` is passed as the value of the _Modal's_ `focusOnMount` _prop_, then we force the hook to focus the "first element" in the element to which the `ref` returned from the hook is applied. 

The default for the hook is `firstElement` but that could change and so I prefer to be absolutely explicit in the behaviour we want and expect rather than rely on defaults. For this reason I map `firstContentElement` to `firstElement`.

You will not that in the Modal's rendering code, we conditionally apply the `ref` returned from the `focusOnMount` hook to different target JSX elements based on the value of the Modal's `focusOnMount` _prop_. If `firstContentElement` is passed then the consumer wishes to place focus on the first element within the Modal's _content_ and thus we apply the `ref` to that element. Otherwise we default to previous behaviour.

I hope that helps to explain this unusual code? I have also provided a comment to further enhance this.

File: packages/components/src/modal/index.tsx
Comment (getdave) [reply packages/components/src/modal/index.tsx]: Ah nice. Apply at the point of use. That's great. Thank you 🙇

**Comment by ciampo** [comment]
Sorry, been a bit busy. Will take a look this afternoon

File: packages/components/src/modal/index.tsx
Comment (draganescu) [reply packages/components/src/modal/index.tsx:90]: This is the gist of the explanation:

> The default for the hook is firstElement but that could change and so I prefer to be absolutely explicit in the behaviour we want and expect rather than rely on defaults. For this reason I map firstContentElement to firstElement.

Thank you!

**Review by draganescu — APPROVED** [review]
Tentative approval.

I tested this and it works great. I also think this implementation is more straightforward compared to #54590 and the whole [tabbable](https://github.com/WordPress/gutenberg/pull/54296/files#diff-7d9147900701280de3fb09897ba9b4e36f424b18412332fc2f72774ef1afe035R55) thing :)

Nevertheless I will defer to someone else more informed from the previous discussion and not merge.

**Comment by getdave** [comment]
I'll await a final approval from @ciampo or @kevin940726 who have been heavily involved in this route 🙇

File: packages/components/src/modal/test/index.tsx
Comment (ciampo) [inline packages/components/src/modal/test/index.tsx]: Can we also add a unit test about `focusOnMount="firstElement"`, like we had in #54296?

File: packages/components/src/modal/test/index.tsx
Comment (ciampo) [inline packages/components/src/modal/test/index.tsx]: No need to use `testid` — we should prefer using roles and text contents when possible

```suggestion
			expect(
				screen.getByRole( 'link', {
					name: 'First Focusable Element',
				} )
			).toHaveFocus();
```

(which means there's no need to assign the `data-testid` attribute to the anchor link above)

File: packages/components/src/modal/test/index.tsx
Comment (ciampo) [inline packages/components/src/modal/test/index.tsx]: We can DRY up the code a little bit by refactoring the React test component outside of the unit tests

<details>

<summary>Something like this</summary>

```diff
diff --git a/packages/components/src/modal/test/index.tsx b/packages/components/src/modal/test/index.tsx
index 0a3111a6df..9baf611034 100644
--- a/packages/components/src/modal/test/index.tsx
+++ b/packages/components/src/modal/test/index.tsx
@@ -13,6 +13,7 @@ import { useState } from '@wordpress/element';
  * Internal dependencies
  */
 import Modal from '../';
+import type { ModalProps } from '../types';
 
 const noop = () => {};
 
@@ -238,6 +239,32 @@ describe( 'Modal', () => {
 	} );
 
 	describe( 'Focus handling', () => {
+		const FocusMountDemo = ( {
+			focusOnMount,
+		}: Pick< ModalProps, 'focusOnMount' > ) => {
+			const [ isShown, setIsShown ] = useState( false );
+			return (
+				<>
+					<button onClick={ () => setIsShown( true ) }>📣</button>
+					{ isShown && (
+						<Modal
+							focusOnMount={ focusOnMount }
+							onRequestClose={ () => setIsShown( false ) }
+						>
+							<p>Modal content</p>
+							<a href="https://wordpress.org">
+								First Focusable Element
+							</a>
+
+							<a href="https://wordpress.org">
+								Another Focusable Element
+							</a>
+						</Modal>
+					) }
+				</>
+			);
+		};
+
 		let originalGetClientRects: () => DOMRectList;
 
 		beforeEach( () => {
@@ -264,34 +291,7 @@ describe( 'Modal', () => {
 		it( 'should focus the first focusable element in the contents (if found) when `firstContentElement` passed as value for `focusOnMount` prop', async () => {
 			const user = userEvent.setup();
 
-			const FocusMountDemo = () => {
-				const [ isShown, setIsShown ] = useState( false );
-				return (
-					<>
-						<button onClick={ () => setIsShown( true ) }>📣</button>
-						{ isShown && (
-							<Modal
-								focusOnMount="firstContentElement"
-								onRequestClose={ () => setIsShown( false ) }
-							>
-								<p>Modal content</p>
-								<a
-									href="https://wordpress.org"
-									data-testid="first-focusable-element"
-								>
-									First Focusable Element
-								</a>
-
-								<a href="https://wordpress.org">
-									Another Focusable Element
-								</a>
-							</Modal>
-						) }
-					</>
-				);
-			};
-
-			render( <FocusMountDemo /> );
+			render( <FocusMountDemo focusOnMount="firstContentElement" /> );
 
 			const opener = screen.getByRole( 'button' );
 
@@ -306,33 +306,8 @@ describe( 'Modal', () => {
 
 		it( 'should focus the Modal dialog when `true` passed as value for `focusOnMount` prop', async () => {
 			const user = userEvent.setup();
-			const FocusMountDemo = () => {
-				const [ isShown, setIsShown ] = useState( false );
-				return (
-					<>
-						<button onClick={ () => setIsShown( true ) }>📣</button>
-						{ isShown && (
-							<Modal
-								focusOnMount
-								onRequestClose={ () => setIsShown( false ) }
-							>
-								<p>Modal content</p>
-								<a
-									href="https://wordpress.org"
-									data-testid="first-focusable-element"
-								>
-									First Focusable Element
-								</a>
 
-								<a href="https://wordpress.org">
-									Another Focusable Element
-								</a>
-							</Modal>
-						) }
-					</>
-				);
-			};
-			render( <FocusMountDemo /> );
+			render( <FocusMountDemo focusOnMount /> );
 
 			const opener = screen.getByRole( 'button' );
 
@@ -343,33 +318,8 @@ describe( 'Modal', () => {
 
 		it( 'should not move focus when `false` passed as value for `focusOnMount` prop', async () => {
 			const user = userEvent.setup();
-			const FocusMountDemo = () => {
-				const [ isShown, setIsShown ] = useState( false );
-				return (
-					<>
-						<button onClick={ () => setIsShown( true ) }>📣</button>
-						{ isShown && (
-							<Modal
-								focusOnMount={ false }
-								onRequestClose={ () => setIsShown( false ) }
-							>
-								<p>Modal content</p>
-								<a
-									href="https://wordpress.org"
-									data-testid="first-focusable-element"
-								>
-									First Focusable Element
-								</a>
 
-								<a href="https://wordpress.org">
-									Another Focusable Element
-								</a>
-							</Modal>
-						) }
-					</>
-				);
-			};
-			render( <FocusMountDemo /> );
+			render( <FocusMountDemo focusOnMount={ false } /> );
 
 			const opener = screen.getByRole( 'button' );
 
```
</details>

File: packages/components/src/modal/index.tsx
Comment (ciampo) [reply packages/components/src/modal/index.tsx]: I was also going to suggest using `useMergeRefs`, thank you @stokesman !

**Review by ciampo — APPROVED** [review]
Code changes LGTM, and test well as advertised. IMO this is a cleaner approach than #54296 (thank you @kevin940726 for the suggestion!).

I left a few comments aimed at improving unit tests — feel free to merge once they have been addressed 🚀 

Thank you!

**Comment by ciampo** [comment]
(also, I don't think this PR needs a dev note?)

**Comment by getdave** [comment]
No need for the dev note on this PR 👍

**Comment by andrewhayward** [comment]
Currently, `firstContentElement` acts like `false` if there are no focusable content elements. Is that expected? I thought we'd discussed in #54296 falling back to `firstElement` behaviour if the only focusable elements were in the header, but I'm not sure if I'm making that up now.

**Comment by stokesman** [comment]
> we'd discussed in https://github.com/WordPress/gutenberg/pull/54296 falling back to firstElement behaviour

Yes, I was wondering if someone would bring that up here as well but I think given that this is a new API (not reusing `firstElement`) it's okay to move forward with this as is. We can add fallback behavior in a follow up if we do believe it to be important. Good point to consider I might be overlooking something.

**Comment by ciampo** [comment]
> > we'd discussed in #54296 falling back to firstElement behaviour
> 
> Yes, I was wondering if someone would bring that up here as well but I think given that this is a new API (not reusing `firstElement`) it's okay to move forward with this as is. We can add fallback behavior in a follow up if we do believe it to be important. Good point to consider I might be overlooking something.

Let's do this, so that this PR at least can get merged ASAP.

File: test/e2e/specs/editor/various/block-renaming.spec.js
Comment (kevin940726) [inline test/e2e/specs/editor/various/block-renaming.spec.js]: Nit: prefer `getByRole` whenever possible 👍

File: packages/components/src/modal/README.md
Comment (kevin940726) [inline packages/components/src/modal/README.md]: Might be worth adding a short sentence here explaining that it's the responsibility of the consumer to add at least one focusable element in `children` or the focus will be lost.

**Review by kevin940726 — APPROVED** [review]
LGTM 👍. Nicely done everyone! ❤️

File: packages/components/src/modal/README.md
Comment (stokesman) [reply packages/components/src/modal/README.md]: Also, is "tabbable" not the more accurate term instead of "focusable"? If not, maybe the preceding sentence should be edited for consistency.

File: packages/components/src/modal/test/index.tsx
Comment (getdave) [reply packages/components/src/modal/test/index.tsx]: ✅ Added test. Good catch 😅

File: packages/components/src/modal/test/index.tsx
Comment (getdave) [reply packages/components/src/modal/test/index.tsx]: ✅ Updated to prefer standard assertions byRole

File: packages/components/src/modal/test/index.tsx
Comment (getdave) [reply packages/components/src/modal/test/index.tsx]: ✅ Refactor complete.

File: test/e2e/specs/editor/various/block-renaming.spec.js
Comment (getdave) [reply test/e2e/specs/editor/various/block-renaming.spec.js]: ✅

File: packages/components/src/modal/README.md
Comment (getdave) [reply packages/components/src/modal/README.md]: ✅ Reworded and edited for consistency.

**Comment by getdave** [comment]
> firstElement

No you're not making it up 😄 Good point to raise.

My opinion - if you opt to use `firstContentElement` then you _must_ ensure there is at least one tabbable element. This is now clarified in the updated `README`. I agree with the other comments that we should identify the need for this "fallback" feature before we added it here.

**Comment by getdave** [comment]
I've set the PR to merge. We can follow up with any future changes.

Thanks to everyone here who contributed and helped get this over the line. I really appreciate the level of detail you all helped foster here 🙇

File: packages/components/src/modal/test/index.tsx
Comment (ciampo) [inline packages/components/src/modal/test/index.tsx]: Removing `focusOnMount` prop as per test description

```suggestion
			render( <FocusMountDemo /> );
```

Comment (getdave) [reply packages/components/src/modal/test/index.tsx]: How did I miss this? 🤦

**Comment by ciampo** [comment]
> I agree with the other comments that we should identify the need for this "fallback" feature before we added it here.

IMO, I think it's still worth following up with a PR introducing a fail-safe fallback behaviour, since it just brings peace of mind to consumers of the modal.

For example, let's say that:

- dev A today adds a `<Modal focusOnMount="firstContentElement" />`, ensuring that there is a focusable element in the modal's contents
- in a year's time, dev B needs to make a quick change which involves removing that focusable element from the modal. Unless another person spots the lack of focus while testing, that change would introduce an unwanted a11y regression.

I'm not saying that this is super urgent, but we should still plan on introducing this fallback behaviour sooner rather than later.

**Comment by getdave** [comment]
It's going to be pretty challenging with the current implementation though as it relies on moving the focus ref between elements.

Another option could be to `console.warn()` if no focusable is found?

**Comment by ciampo** [comment]
> It's going to be pretty challenging with the current implementation though as it relies on moving the focus ref between elements.
> 
> Another option could be to `console.warn()` if no focusable is found?

A `console.warn()` is probably better than nothing, but I'd personally like to find a way to fallback more gracefully. We can always rework the implementation if necessary, especially now that we have a separate value for this behaviour.

**Comment by getdave** [comment]
Pinging @mikachan and @priethor so we can get this landed in WP 6.4 🙏

**Comment by afercia** [comment]
Late to the party, sorry. I'm not sure this is going in the right direction.

> As discussed in #54106 dialogs (including Modal) should ideally place focus on the first focusable element on mount.

This is a common misconception and, in a way, now considered an 'old' recommendation. I've been closely following the discussion about initial focus for the native `<dialog>` element in the whatwg / html GitHub at https://github.com/whatwg/html/pull/4184

It is very important to note that after long discussion the initial focus mechanism will be changed. If no specified otherwise, the `<dialog>` element itself will receive initial focus. Browser makers are alredy working on implementhing the new focus algorithm, and some may already have implemented it.

Although the above relates to the native HTML `<dialog>` element, I think our component should match that behavior and set initial focus on the dialog element itself, if not specified otherwise.

Cc @joedolson @alexsanford @andrewhayward 

Aside:
The component name `Modal` is really inaccurate. Although commonly used to identify an user interface pattern, the term 'modal' is actually a _behavior_. Many user interface patterns other than a dialog may have a modal behavior. This should have been named `Dialog` or, better, `ModalDialog`, as dialogs can be modal and also non-modal.

**Comment by ciampo** [comment]
> It is very important to note that after long discussion the initial focus mechanism will be changed. If no specified otherwise, the `<dialog>` element itself will receive initial focus. Browser makers are alredy working on implementhing the new focus algorithm, and some may already have implemented it.
> 
> Although the above relates to the native HTML `<dialog>` element, I think our component should match that behavior and set initial focus on the dialog element itself, if not specified otherwise.

@afercia , The `Modal` component already focuses the `div[role="dialog"]` as its default behaviour (see [storybook](https://wordpress.github.io/gutenberg/?path=/story/components-modal--default)). This PR doesn't change any of that, and only adds one additional behaviour that consumers of the component can choose from.

> The component name `Modal` is really inaccurate. Although commonly used to identify an user interface pattern, the term 'modal' is actually a _behavior_. Many user interface patterns other than a dialog may have a modal behavior. This should have been named `Dialog` or, better, `ModalDialog`, as dialogs can be modal and also non-modal.

Thank you for the advice. The `Modal` component has been around for a very long time, and I'm not sure what the reasons to call it that way were at the time. As we're soon going to start refactoring the component to use `ariakit` under the hood, we'll be also considering the component naming.

**Comment by mikachan** [comment]
I just cherry-picked this PR to the release/16.7 branch to get it included in the next release: 76e521409a

**Comment by afercia** [comment]
> Thank you for the advice. The `Modal` component has been around for a very long time, and I'm not sure what where the reasons to call it that way.

Lack of knowledge and uncritically adoption of (long established) wrong naming, i guess.

**Comment by ciampo** [comment]
Added a dev note in the PR description

**Pull request merged**
